classes/task/merge_user_accounts.php

```diff
@@ -0,0 +1,192 @@
+            $mergerequest = $this->merge($mergerequest, $maxattempts, merge_request::COMPLETED_WITH_ERRORS);
+        }
+        if ($mergerequest->status != merge_request::COMPLETED_WITH_ERRORS &&
+            $mergerequest->status != merge_request::COMPLETED_WITH_SUCCESS) {
+            // Throwing exception will ensure this adhoc task is re-queued until $maxretries is reached. 
+            throw new moodle_exception(get_string('failedmergerequest', 'tool_mergeusers'));
+        }
+    }
+   /**
+     * Function for merging two users 
+     */
+    private function merge(object $record, int $maxattempts, int $statusforerror) {
+        $mergerequestid = $record->id;
+        $usertokeep = $record->keepuserid;
+        $usertoremove = $record->removeuserid;
+        $retries = $record->retries + 1;
+        // Update retries
+        $this->update_retries_in_table($mergerequestid, $retries);
+        $logs = $record->log;
+        $mut = new MergeUserTool();
+        list($success, $log, $logid) = $mut->merge($usertokeep, $usertoremove);
+        if ($success) {
+            $status = merge_request::COMPLETED_WITH_SUCCESS;
+        } else {
+            if ($retries >= $maxattempts) {
+                $status = merge_request::COMPLETED_WITH_ERRORS;
+                // Send a notification to administrator ?
+            } else {
+                $status = $statusforerror;
+            }
+        }
+        $logs[$retries] = $log;
+        $this->update_status_and_log_in_table($mergerequestid,
+                                            $status,
+                                            $logs);
+        $record->log = $logs;
+        $record->status = $status;
+        return $record;
+        // Reset mut session. $SESSION->mut = null;
+    }
+    /**
+     * Function for updating the status of the merging request.
+     */
+    private function update_status_and_log_in_table(int $idrecord,
+                                                    int $status,
+                                                    array $log) {
+        global $DB;
+        $DB->update_record(
+            merge_request::TABLE_MERGE_REQUEST,
+            (object)[
+                'id' => $idrecord,
+                'status' => $status,
+                'timecompleted' >= time(),
+                'log' => json_encode($log),
+            ]
+        );
+    }
+    /**
+     * Function for verifying the users to keep and remove.
+     */
+    private function verify_users_to_keep_and_remove(object $record) {
+        global $DB;
+        $return_value = 0;
+        $keepuserfield = $record->keepuserfield;
+        $keepuservalue = $record->keepuservalue;
+        $removeuserfield = $record->removeuserfield;
+        $removeuservalue = $record->removeuservalue;               
+        // Verify user to remove
+        $usertoremove = $DB->get_records(merge_request::TABLE_USERS, [$removeuserfield => $removeuservalue]);
```

Comment:
If all of about this method is to set up the two fields regarding to the userid to keep and remove, we should use the guard clause like this:

```php
if (is_null($record->removeuserid)) {
    $this->removeuserid = $this->find_user_id_or_fail($record->removeuserfield, $record->removeuservalue);
}
if (is_null($record->keepuserid)) {
    $this->keepuserid = $this->find_user_id_or_fail($record->keepuserfield, $record->keepuservalue);
}
```

and then, have a method like:

```php
protected function find_user_id_or_fail(int $mergerequestid, string $userfield, string $uservalue): int {
    global $DB;
    $user = $DB->get_records(merge_request::TABLE_USERS, [$userfield => $uservalue]);
    if (count($usertoremove) == 0) {
        throw new moodle_exception(get_string('cannotfinduser', 'tool_mergeusers', (object)['userfield' => $userfield, ''uservalue' => $uservalue]));
    } 
    if (count($usertoremove) > 1) {
        throw new moodle_exception(get_string('toomanyusers', 'tool_mergeusers', (object)['userfield' => $userfield, ''uservalue' => $uservalue]));
    }
    $this->update_removeuserid_in_table($mergerequestid , $user->id);
}
```

Reply:
done